sensor/DeviceServiceScanSensor.js

```diff
@@ -15,138 +15,250 @@
+                                    hosts: tagHosts,
+                                    policy: policy
+                                })
+                            } else {
+                                this.tagScanMap[tagUid] = {
+                                    hosts: tagHosts,
+                                    policy: policy
+                                }
+                            }
 
-    return null;
-  }
+                            await this.applyDeviceScan(this.tagScanMap[tagUid]);
+                        }
+                        break;
+                    }
+                    case "NetworkProfile": {
+                        const uuid = host.o && host.o.uuid;
+                        if (uuid) {
+                            const hosts = await hostManager.getHostsAsync();
+                            const networkHosts = hosts.filter((h) => {
+                                return h && h.o && h.o.intf_uuid == uuid
+                            })
+                            if (this.networkScanMap[uuid]) {
+                                this.networkScanMap[uuid] = Object.assign(this.networkScanMap[uuid], {
+                                    hosts: networkHosts,
+                                    policy: policy
+                                })
+                            } else {
+                                this.networkScanMap[uuid] = {
+                                    hosts: networkHosts,
+                                    policy: policy
+                                }
+                            }
+                            await this.applyDeviceScan(this.networkScanMap[uuid]);
+                        }
+                        break;
+                    }
+                    case "Host": {
+                        const macAddress = host && host.o && host.o.mac;
+                        if (macAddress) {
+                            if (this.deviceScanMap[macAddress]) {
+                                this.deviceScanMap[macAddress] = Object.assign(this.deviceScanMap[macAddress], {
+                                    hosts: [host],
+                                    policy: policy
+                                })
+                            } else {
+                                this.deviceScanMap[macAddress] = {
+                                    hosts: [host],
+                                    policy: policy
+                                }
+                            }
+                            await this.applyDeviceScan(this.deviceScanMap[macAddress]);
+                        }
+                        break;
+                    }
+                    default:
+                }
+            }
+        } catch (err) {
+            log.error("Got error when applying adblock policy", err);
+        }
+    }
 
-  async runOnce() {
-    log.info('Scan start...');
+    async applyScan() {
```

Comment:
this is a little bit too complex form what i can see.

Can we have a global list of all devices, and mark them as if them should be scanned on each iteration, by system -> network -> group -> device policy? Device policy comes at last and have the highest priority.

And there should be only one interval timer to control the scan. which is controlled by the global switch.

Reply:
I thought about this way, but the policy may be async, we need to control them by different level.

Reply:
e.g. 
if the device1 port scan is disabled, but tag1 port scan is enabled. will clear device1 scan timer and create a scan job for tag1, tag1 may include device1. when box restart, we may not confirm which policy will be applied first...

Reply:
go through every policy on each iteration, port scan is a heavier job than reading redis. Is it even possible in the current design to exclude a device from scanning in a network?